### PR TITLE
[Bugfix][TENT] Predict deadline MLU from a metered wire rate and the bytes ahead

### LINKED ISSUE #3750: [Bug]: TENT deadline predictors disagree on bandwidth, and the drop predictor multiplies worker-queue wait by the request's slice count
### Bug Report

**Environment:** `main` (observed at 69c22878), TENT RDMA transport with the runtime queue enabled. No hardware needed to reproduce the arithmetic — it is in the code paths below.

**Summary**

A request carrying `deadline_ns` is judged by its *predicted MLU* (predicted completion time / remaining window). Two layers make a decision from that number, and each computes it from a different bandwidth through its own copy of the formula:

| Layer | Decision | Bandwidth used |
|---|---|---|
| `LocalTransferAdmissionQueue::pickForDispatch` → `shouldDrop` (RFC #2519 step 3, #2764/#2816) | drops the owner → `CANCELED` | `RdmaTransport::getEstimatedBandwidth()` = sum of every device's **selection EWMA** (`DeviceSelector::getAggregateEwmaBandwidth`) |
| `Workers::asyncPostSend` deadline arbitration (RFC #2792, #2794) | orders QP posting | the constant `transports/rdma/default_bandwidth_gbps` |

So the same request is ranked on two unrelated scales, and the layer that can cancel it stands on the weaker number. Three concrete defects:

**1. The selection EWMA folds worker-queue wait into "bandwidth", and the drop predictor then multiplies that wait by the request's slice count.**

`Workers::asyncPollCq` feeds `DeviceSelector::release()` with `overall_lat = poll_ts - enqueue_ts` per slice, i.e. queue wait `q` plus wire time. For a slice of length `l` on a link of rate `r` the learned "bandwidth" is `l / (q + l/r)`. `shouldDrop` predicts a request of `L` bytes as `L / bw`:

```
L / bw = L · (q + l/r) / l = (L/l) · q + L/r
```

The wait is charged once per slice, but a request's slices wait behind the same backlog concurrently — they wait once. A 64-slice request is predicted to wait 64×. Large requests are therefore systematically over-estimated and preferentially dropped, and the admission queue's own dispatch window (which is what fills the worker queue) feeds back into its drop decision.

**2. The series is ~99% the latest sample.** `bandwidth_learning_rate` defaults to 0.01 and weights the *old* value (`new = α·old + (1-α)·observed`), so a single slow completion can flip an irreversible drop.

**3. The arbitration's constant was never measured**, and after #3644 (per-NIC negotiated speed) it still reads the config default rather than the device's rate.

The selection EWMA cannot simply be "fixed": it includes queueing on purpose so that a backed-up NIC scores worse for device selection.

**What RFC #2519 actually specified**

> `available_bw` — derived from the existing `ewma_bandwidth_bps` **minus current in-flight load** on the pinned device.

The in-flight term was dropped when the predictor landed (#2764, #2816). The bug above is what happens when queueing is instead smuggled into the rate.

**Expected**

Both layers predict completion time the way `DeviceSelector` already scores devices — `(bytes_ahead + length) / wire_rate` — from one definition, with `bytes_ahead` as an additive term (what the request must wait behind: bytes already dispatched and not yet completed) over a queue-free wire rate, smoothed heavily enough that one outlier cannot cancel a user's transfer.

**Fix**

Prepared on `SongOf:fix/tent-mlu-bandwidth-estimate`, stacked on #3644 (it needs the per-device link speed as the seed and the availability flag for the aggregate):

- one predictor, `DeadlineMlu(bytes_ahead, length, deadline_ns, now_ns, bw_bps)` in `tent/runtime/deadline_mlu.h`, called by both `shouldDrop` and `PredictedMlu`;
- a per-device **transmit estimate** (`ewma_transmit_bps`): same update rule, clamp and link-speed seed as the selection EWMA, but fed post→completion time only and smoothed with its own `transmit_bandwidth_learning_rate` (default 0.99); the selection EWMA is unchanged;
- `bytes_ahead` = the admission queue's `dispatching_bytes_` (+ what the current pick already took) for the drop, and the NIC's inflight bytes for the arbitration;
- `getEstimatedBandwidth()` returns the aggregate transmit estimate; the arbitration reads the local NIC's.

Tests reproduce each point without hardware (same samples → selection EWMA 1.55 GB/s vs transmit 3.1 GB/s; a 30 GB/s outlier moves the selection EWMA above 25 GB/s but the transmit estimate by ~1%; a 16 B / 500 µs-window owner is dispatched on an idle queue, dropped behind a 1 MB owner in flight, and dispatched again once it completes). PR to follow once #3644 merges.

Refs #2519 (restores the in-flight term), #2764, #2816, #2794, #2132.

---

### Follow-up left open by #3644 — addressed by #3777

The predictor above divides by a per-NIC bandwidth that #3644 seeds from the negotiated port speed and re-reads on port events. Its review left one input out of scope:

**LAG-effective speed (`ibv_query_port_speed()`).** `linkSpeedGbps()` decoded `active_speed × active_width`, the physical link's encoded rate. For a VF over VF LAG / multiport LAG, a PF dropping out of the bond halves the VF's effective bandwidth while its port stays `ACTIVE` at the same encoding, so the re-read on `IBV_EVENT_DEVICE_SPEED_CHANGE` saw `after == before` and did not re-seed. rdma-core v62 added `ibv_query_port_speed()` alongside the event to report the *effective* speed (100 Mb/s units).

**#3777** resolves it as an optional `IbvLoader` symbol (older libraries keep RDMA and the encoded rate), prefers it in `linkSpeedGbps()` via `ibPortSpeedGbps()`, and adds a verbs-injection seam on `RdmaContext` so the effective-speed and refresh paths are tested without an RNIC.

### Before submitting...

- [x] Ensure you searched for relevant issues and read the documentation

### PR DESCRIPTION
**Pull request by SongOf**
## Description

Fixes #3750.

The admission queue's deadline-infeasible drop (RFC #2519 step 3) and the RDMA workers' bandwidth arbitration (RFC #2792) each computed "predicted MLU" from a different bandwidth: the drop from the sum of the selection EWMAs, the arbitration from the constant `default_bandwidth_gbps`. The selection EWMA is fed per-slice latency, so it folds queueing into a "bandwidth"; predicting an `N`-slice request from it multiplies that wait by `N` although the slices wait behind the same backlog concurrently, and one slow completion could flip an irreversible drop.

Both layers now predict completion time as `(bytes_ahead + length) / wire_rate` through one definition, `DeadlineMlu` (`tent/runtime/deadline_mlu.h`), restoring the in-flight term RFC #2519 specified. Review of the first version (thanks @alogfans, @catyans) found that neither input was what it claimed to be, and fixing that reached into the per-slice accounting. Three commits, each building and passing on its own:

1. `[Bugfix][TENT] Return every slice's charge and lane count to the counter that holds it` — the per-slice accounting below (retry re-charge, timeout release, shared-QP ownership), nothing the predictors read yet.
2. `[Bugfix][TENT] Meter the NIC's transmit rate and predict the admission drop from it and the bytes ahead` — `DeadlineMlu`, the transmit meter, `dispatching_bytes_`, `getEstimatedBandwidth()`.
3. `[TENT] Score each arbitration slot against the bytes actually posted ahead of it` — the arbitration reads the meter and the posted backlog, greedily per slot.

**Wire rate.** Each device gains a transmit estimate, `ewma_transmit_bps`, with the selection EWMA's update rule, clamp and link-speed seed, but fed from a **meter**: bytes completed over the time the NIC spent with work posted (`posted_bytes` / `busy_ns` / `busy_since`, `maybeSampleTransmit`), one sample per `transmit_meter_interval_ns` (10 ms), smoothed with `transmit_bandwidth_learning_rate` (0.9 ≈ 10 intervals). Per-completion latency cannot serve: work requests are posted in batches whose timestamps are effectively one and a poll pass stamps its completions alike, so a slice's own post→completion grows with batch depth (R, R/2, R/3, …). Busy time rather than elapsed time, because a workload that bursts and waits otherwise reads its duty cycle as the link rate; the sample is taken only on the last completion of a poll pass; a posted slice that ends without moving its bytes (failed, flushed, timed out) resets the interval. `getEstimatedBandwidth()` returns the aggregate transmit estimate; the arbitration reads the local NIC's. The selection EWMA is main's (#3511): a successful attempt's post→completion time.

**Bytes ahead.** The admission queue charges the bytes of drop-eligible owners it has dispatched and not yet completed (`dispatching_bytes_`, plus what the current pick already took); owners on other transports share the queue but not the NIC. The arbitration reads the NIC's **posted bytes** — what has reached the hardware — not the selector's allocation-time `inflight_bytes`, which already held the slices being ordered; `OrderByUrgency` then scores each slot behind the slots chosen before it (greedy for the first 64, one ranked pass for the rest).

**Per-slice accounting.** The meter and the arbitration only mean something if every charge comes back to the counter it was made on, and with `qp_pools` several worker lanes share a queue pair, so a completion or a timeout sweep on one lane routinely pops slices another lane enqueued. Every accounting a slice carries now names the counter it sits in — `charged_dev`, `posted_dev`, `counted_lane` (atomic, −1 = none) — and whichever path takes it out exchanges the field for −1 and pays back exactly what it read, so no path pays twice, none pays the wrong lane or device, and a fallback rewriting `source_dev_id` or a retry moving between lanes cannot misdirect a release. `owner_worker` routes set entries: a sweeper hands another lane's entry to that lane's reclaim list, drained on its next pass together with entries a teardown left behind. Retries are re-charged on the device they actually post on (`rechargeSlice`, charge-then-swap); a timeout returns the charge of the slice and of everything swept with it (`acknowledge` gained a per-slice callback); a slice resolved while it waited for a re-post is dropped rather than posted; everything a shared QP's poller must find (post timestamp, posted device, set entry) is written inside `submitSlices` before the WR reaches the wire. `selectMultiPath` charges each slice its real length, so release by length balances without a recorded charge size. `DeviceInfo` is 64-byte aligned with its hot counters on separate lines.

**Relation to #3511** (merged while this was in review): it fixed the retry/fallback charge with `charged_dev_id` + `charged_bytes` + a release-then-charge `chargeSliceQuota`. This PR keeps its own mechanism (the exchange discipline above is what a pooled QP's races need, and exact charging removes the need for `charged_bytes`) and follows #3511's semantics on the selection-EWMA sample and on baseline mode tracking no inflight. `quota_accounting_test.cpp` keeps #3511's selector-level cases; its slice-level cases live in `RdmaWorkersChargeTest`.

Known and deliberately left out: `submitSlices` and `acknowledge` hold only the endpoint's read guard, so on a pooled layout two lanes can reach one `BoundedSliceQueue` at once. That race predates this PR and needs a per-QP lock (noted in the `slice_queue_` comment); separate change.

No public API or wire-format changes. New config keys: `transports/rdma/transmit_bandwidth_learning_rate`, `transmit_meter_interval_ns`, `transmit_meter_max_interval_ns`.

## Module

- [x] Transfer Engine (`mooncake-transfer-engine`)
- [x] Docs

## Type of Change

- [x] Bug fix

## How Has This Been Tested?

**Test commands:**
```bash
cd build
cmake --build . -j4
ctest -R 'tent_|admission_queue|bw_arbitration' --output-on-failure
```

**Test results:**
- [x] Unit tests pass — 47/47 in the selection above; full `ctest` 94/98, the 4 failures need NUMA sysfs / a network the container lacks and fail identically on `main` (Ubuntu 22.04 container, rdma-core 39, no RNIC)
- [ ] Integration tests pass (if applicable)
- [x] Manual testing done (describe below)

All new tests run without hardware. Those that pin a specific line were checked by flipping it (the test then fails).

- `deadline_mlu` / `bw_arbitration_test`: `DeadlineMlu` edge cases; each greedy slot is scored behind the ones before it; bytes ahead favour the tighter window.
- `admission_queue_test`: an owner is dispatched on an idle queue, dropped behind 1 MB of eligible work in flight, dispatched again once it completes; other-transport owners do not count; no bandwidth / no deadline never drops.
- `tent_device_selector_test`: the estimate is independent of batch depth (1..64 slices at one timestamp) and slice size; busy time counts only stretches with work posted; idle gaps and stale intervals are not samples; a reset re-baselines. `DeviceSelectorTransmitAccuracyTest` pins the number over ten successive stretches: a steady wire exact, a bursty one within 0.4 %, a duty cycle to 1e-6, a 4x slowdown followed within 478 ms.
- `tent_quota_accounting_test` (#3511's, re-pointed at the selector API): per-slice charge/release balance incl. a short trailing slice, `chargeDevice`, baseline mode keeps inflight at zero, EWMA gating.
- `tent_rdma_transport_test`:
  - `RdmaWorkersArbitrationTest`: `orderByDeadline` with a live selector — allocated contenders are not counted as queue-ahead, posted work outside the batch is.
  - `RdmaWorkersChargeTest`: a retry is re-charged and its completion teaches the selector; a fallback moves the charge; a charge on an untracked device keeps the old one.
  - `RdmaWorkersCompletionTest` (synthesized `ibv_wc` through `handleCompletion`): success feeds the meter; idle between bursts is not charged; a failed slice ends its stretch without teaching; only the last completion of a pass samples; a resolved slice still leaves the lane count; a COMPLETED sweep keeps the interval, a FAILED one abandons it.
  - `RdmaWorkersSharedQpTest` (a real `RdmaEndPoint` on fake verbs, one QP shared by two lanes): a timeout sweep on lane 0 gives lane 1 its slice back with both sets, counters and charges balanced; teardown leftovers are reclaimed by their lane; a retried slice stays counted and its completion is a transmit sample; a retry polled by the other lane moves ownership; a stale set entry is taken out by the lane holding it; a slice resolved while queued is dropped, not posted; a completion sweep keeps the meter interval.
  - `RdmaWorkersTimeoutTest`: a timed-out slice returns its charge with its endpoint gone or no longer holding it.

## Checklist

- [x] I have performed a self-review of my own code
- [x] I have formatted my code using `./scripts/code_format.sh`
- [x] I have run pre-commit on the files changed in this PR and all hooks pass
- [x] I have updated the documentation (if applicable)
- [x] I have added tests to prove my changes are effective
- [ ] For changes >500 LOC: I have filed an RFC issue (~1.1k LOC excluding tests; #3750 describes the change and the review comments above the rest — happy to file an RFC if wanted)

## AI Assistance Disclosure

- [ ] No AI tools were used
- [x] AI tools were used (specify below)

Claude Code was used for drafting code, tests and comments; every changed line was reviewed by the submitter.

### CODE REVIEW

**Comment by catyans** [comment]
The new arbitration `bytes_ahead` is not actually the number of bytes ahead of the candidates being ordered. `DeviceSelector::allocate()` charges every slice before it enters the worker queues, so by the time `asyncPostSend()` calls `getInflightBytes()`, that value already includes every slice in the current `slices` vector (and may include work queued behind it). Passing the full value to `PredictedMlu()` and then adding `flow.length` counts the current slice twice and treats all of its siblings as preceding it. This differs from the admission model, where `dispatching_bytes_` excludes the owner currently being evaluated and grows only after an earlier owner is selected.

There is a complementary undercount on retry: the CQ/error paths call `releaseSliceQuota()` before `submit(slice)`, which clears `quota_charged`, but `source_dev_id` remains non-negative. `selectOptimalDevice()` therefore skips `DeviceSelector::allocate()` on the retry, so the retry is never charged again; its eventual completion also cannot feed either EWMA because `releaseSliceQuota()` sees the flag already clear.

Please make the queue-ahead accounting represent only work that actually precedes a candidate, and re-establish the selector charge for each retried attempt. At minimum, arbitration should exclude the current candidate set from the pre-existing inflight term and then account for ordering explicitly. A production-path regression test should allocate multiple differently sized slices, pass them through worker arbitration, and verify each computed queue-ahead value; a second test should fail and retry a charged slice, then verify inflight remains balanced and the successful retry contributes a transmit sample.

File: mooncake-transfer-engine/tent/src/transport/rdma/workers.cpp [hunk 1]
```diff
@@ -684,7 +719,13 @@ void Workers::asyncPollCq() {
                 (slice->submit_ts - slice->enqueue_ts) / 1000.0;
             double inflight_lat = (poll_ts - slice->submit_ts) / 1000.0;
             double overall_lat_sec = (poll_ts - slice->enqueue_ts) / 1e9;
-            releaseSliceQuota(slice, overall_lat_sec);
+            // Only a successful completion measures wire time; a burst of
+            // flush completions would drag the transmit estimate down for
+            // hundreds of completions.
+            double transmit_lat_sec = wc[i].status == IBV_WC_SUCCESS
+                                          ? (poll_ts - slice->submit_ts) / 1e9
+                                          : 0.0;
```
Comment (alogfans) [inline mooncake-transfer-engine/tent/src/transport/rdma/workers.cpp]: **Blocking: the transmit EWMA still includes queueing within a posted batch**

`transmit_latency = poll_ts - submit_ts` is not a wire-only service-time sample when several WRs are posted together. Their `submit_ts` values are nearly identical, while completions arrive serially, so equal-sized WRs produce samples approximately `R, R/2, R/3, ...`. CQ polling delay is included as well.

Consequently, the learned rate decreases with batch depth and can converge near the lower clamp even on a healthy link. Since `bytes_ahead` is then added separately, this also counts part of the NIC/QP queueing twice and may cause false, irreversible deadline drops.

Could we instead derive the estimate from device-level completed bytes over elapsed time, or from non-overlapping service intervals? A regression test should post/model multiple equal-sized slices at the same timestamp and verify that the estimate remains approximately independent of batch size.

File: mooncake-transfer-engine/tent/src/transport/rdma/workers.cpp [hunk 2]
```diff
@@ -639,34 +655,53 @@ void Workers::promoteTimedOutRequests(WorkerContext& worker) {
...
             worker.inflight_slices.fetch_sub(num_slices);
-            slice_to_remove.push_back(slice);
+            if (num_slices == 0) {
+                // No longer on the QP (a neighbour's retry popped it, with
+                // its count): nothing swept, so fail it here.
+                releaseSliceQuota(slice);
+                updateSliceStatus(slice, TIMEOUT);
+                slice_to_remove.push_back(slice);
+            }
         }
```
Comment (alogfans) [inline mooncake-transfer-engine/tent/src/transport/rdma/workers.cpp:900]: **Blocking: timeout cleanup is still attributed to the wrong worker with pooled QPs**

`RdmaEndPoint::acknowledge()` may sweep slices belonging to multiple workers when they share a pooled QP. However, `expireTimedOutSlices(worker, ...)` adds every swept slice to the current worker’s `slice_to_remove` and subtracts the entire returned count from that worker’s `inflight_slices`.

A swept slice owned by another `WorkerContext` can therefore remain in its owner’s `inflight_slice_set`, while the current worker’s counter is over-decremented. If the QP is destroyed before its CQEs are polled, the stale entry may never be removed. Making `quota_charged` atomic prevents a duplicate selector release, but it does not fix worker-local ownership/accounting.

Please route cleanup to each slice’s actual owner, for example by recording the owner worker ID and scheduling removal on that worker. This needs a two-worker regression test where both workers share a QP and a timeout on worker A sweeps a slice owned by worker B; both sets, counters, and selector charges should end balanced.

File: mooncake-transfer-engine/tent/src/transport/rdma/workers.cpp [hunk 3]
```diff
@@ -506,13 +514,20 @@ void Workers::asyncPostSend() {
...
-                                      ? device_selector_->getSchedulingParams()
-                                                .default_bandwidth_gbps *
-                                            1e9 / 8.0
+                                      ? device_selector_->getTransmitBandwidth(
+                                            path.local_device_id)
                                       : 0.0;
+            const size_t bytes_ahead =
+                device_selector_
+                    ? device_selector_->getInflightBytes(path.local_device_id)
+                    : 0;
```
Comment (alogfans) [inline mooncake-transfer-engine/tent/src/transport/rdma/workers.cpp]: **Blocking: `getInflightBytes()` does not represent bytes ahead of these contenders**

The selector charges every slice during allocation, before it is posted. Therefore, when `asyncPostSend()` reads `getInflightBytes(local_device_id)`, the value already includes the slices currently being sorted, other unposted slices, and potentially work that will be scheduled after the candidate. `DeadlineMlu()` then adds the candidate length again.

This is not the documented `(bytes_ahead + length) / wire_rate` model and can materially change the ordering because the common overcharge is divided by different deadline windows.

Please track the posted/WQ backlog separately, or subtract the current contenders with well-defined saturating semantics. The test should exercise the live selector-to-worker path rather than passing a synthetic `bytes_ahead` directly to `PredictedMlu()`.

File: mooncake-transfer-engine/tent/src/transport/rdma/workers.cpp (hunk 1 above)
Comment (SongOf) [reply mooncake-transfer-engine/tent/src/transport/rdma/workers.cpp]: Fixed in 80d96aa4 (rebased on main; the branch is now three commits). The per-slice latency sample is gone from the transmit estimate; it is a meter over device-level counters: bytes completed divided by the time the NIC spent with work posted (`DeviceInfo::posted_bytes` / `busy_ns` / `busy_since`, `DeviceSelector::maybeSampleTransmit`). Queueing therefore only enters the prediction once, through `bytes_ahead`.

Two details worth knowing:
- Busy time rather than elapsed time, because a bursty workload otherwise reads its duty cycle as the link rate (16 MiB at 8 GB/s every 10 ms settled at 1.68 GB/s with elapsed time). Busy stretches open/close on the `posted_bytes` 0↔non-zero transitions, so the cost is one timestamp per burst edge.
- The sample is offered only on the last completion of a poll pass (`handleCompletion`'s `last_in_pass`): a pass stamps everything it reaps with one timestamp, and an interval closing partway through a pass paid no time for the rest of it (8% high at 2 GB/s with 16-deep polling).

The *selection* EWMA keeps main's per-completion sample as #3511 defined it (a successful attempt's post→completion time); only the deadline predictors read the meter.

Tests: `DeviceSelectorTransmitTest.RateIsIndependentOfBatchDepth` posts 1..64 equal slices at one timestamp and requires the same estimate for every depth; `SliceSizeDoesNotSkewTheEstimate`, `IdleTimeIsNotChargedToTheLink`, `RdmaWorkersCompletionTest.OnlyTheLastCompletionOfAPassSamples` cover the rest, and `DeviceSelectorTransmitAccuracyTest` pins the number over ten successive stretches (steady wire exact, bursty within 0.4%, 4x slowdown followed in 478 ms).

Residuals: a stretch closes when its last completion is *polled*, so that one CQE's polling delay is inside the busy time; and after a failure/timeout the meter restarts its interval (`resetTransmitMeter`), so the first sample after it only re-baselines — a link learns again from the second completion on.

File: mooncake-transfer-engine/tent/src/transport/rdma/workers.cpp (hunk 2 above)
Comment (SongOf) [reply mooncake-transfer-engine/tent/src/transport/rdma/workers.cpp:900]: Fixed in dd161d67. Cleanup is routed per slice now; nothing is bulk-subtracted from the sweeping worker any more.

- Every accounting a slice carries names the counter it sits in: `counted_lane` (which lane's `inflight_slices`), `charged_dev`, `posted_dev`; -1 = none. Whichever path takes the slice out — its own completion, another lane's sweep of the shared QP, the timeout pass — exchanges the field for -1 and pays back exactly what it read; a second path reads -1 and does nothing. A retry that lands on another lane moves the count in one exchange (`submitFromTick`).
- `owner_worker` names the lane whose `inflight_slice_set` holds the entry. A sweeper that pops someone else's slice hands it to that lane's `reclaimed` list (`routeSetRemoval`), and the owner erases it on its next pass (`drainReclaimed`). Entries left behind when a QP is destroyed before its CQEs are polled are retired by the owner's pass as well (`expireTimedOutSlices` retires non-PENDING leftovers).

On the rebase this replaces #3511's `charged_dev_id`/`charged_bytes`/`chargeSliceQuota`: that reconcile releases and then re-charges non-atomically, which is exactly the window a pooled QP's poller can hit, and with `selectMultiPath` charging each slice its real length there is nothing left for `charged_bytes` to correct. `quota_accounting_test.cpp` keeps #3511's selector-level cases (per-slice balance, baseline mode tracks no inflight, EWMA gating, charge on an untracked device); the slice-level ones (fallback migration, retry re-charge, failed charge) moved to `RdmaWorkersChargeTest`, where a `Workers` is built without hardware.

Tests (`RdmaWorkersSharedQpTest`, no hardware: a real `RdmaEndPoint` constructed on fake verbs, one QP shared by two lanes): `TimeoutSweepGivesTheOtherLaneItsSliceBack` has lane 0's timeout sweep lane 1's slice and asserts both sets, both counters and the selector's inflight/posted bytes end at zero; `TeardownLeftoversAreReclaimedByTheirLane`, `RetryPolledByAnotherLaneMovesOwnership`, `StaleSetEntryIsTakenOutByTheLaneHoldingIt`, `SliceResolvedWhileQueuedIsDroppedNotPosted` cover the other orderings.

One thing this does not fix, on purpose: `submitSlices` and `acknowledge` hold only the endpoint's read guard, so on a pooled layout two lanes can reach one `BoundedSliceQueue` concurrently. That is a data race in the queue itself and needs a per-QP lock; I've left it out of this PR (noted in the `slice_queue_` comment) and will do it separately.

File: mooncake-transfer-engine/tent/src/transport/rdma/workers.cpp (hunk 3 above)
Comment (SongOf) [reply mooncake-transfer-engine/tent/src/transport/rdma/workers.cpp]: Fixed in 23d36017 (the posted backlog itself comes with 80d96aa4). The posted backlog is tracked separately: `DeviceInfo::posted_bytes` is charged when a work request reaches the hardware (`Workers::markPosted`, inside `submitSlices` before `ibv_post_send`) and returned when the slice leaves the QP. `orderByDeadline` reads `getPostedBytes()` as `bytes_ahead` — none of the contenders is in it, nor is anything still sitting in a worker queue — and the greedy pass then adds each chosen slice's length before scoring the next slot, so the model is literally `(bytes_ahead + length) / rate` per slot.

Tests go through the worker's own ordering function with a live `DeviceSelector` rather than a synthetic `bytes_ahead`: `RdmaWorkersArbitrationTest.QueueAheadExcludesTheSlicesBeingOrdered` allocates the contenders (so `getInflightBytes` holds them) and checks the order is the one for zero queue-ahead; `QueueAheadCountsWorkOutsideTheBatch` posts unrelated work first and checks the order flips the way the extra bytes predict.

**Comment by SongOf** [comment]
@catyans both halves are fixed: queue-ahead in 23d36017, the retry re-charge in dd161d67 (rebased on main, which picked up #3511 in the meantime — see the note at the end).

**Queue ahead.** Arbitration no longer reads the allocation-time charge. `bytes_ahead` is the device's posted backlog (`getPostedBytes`, charged when a WR reaches the hardware, returned when it leaves the QP), which by construction excludes the candidate set and anything still queued behind it. Ordering is then explicit: the greedy pass scores each slot against `bytes_ahead` plus the lengths of the slots chosen before it, the same `DeadlineMlu` the admission path uses with `dispatching_bytes_`. `RdmaWorkersArbitrationTest.QueueAheadExcludesTheSlicesBeingOrdered` / `QueueAheadCountsWorkOutsideTheBatch` drive differently sized allocated slices through `Workers::orderByDeadline` with a live selector; `BwArbitrationTest.EachSlotIsScoredBehindTheOnesBeforeIt` pins the per-slot accumulation. They assert the resulting order (the comments give the MLUs each slot must have been scored with), not the intermediate values.

**Retry.** Every retry (`retry_count > 0`) goes through `selectFallbackDevice`, which now calls `rechargeSlice` for the device the attempt will actually use, so the charge is re-established each attempt; the release on completion pays back the device recorded in `charged_dev` rather than whatever `source_dev_id` says by then. The transmit estimate itself no longer depends on the charge at all — it is metered from posted bytes over busy time — so a retry's completion is a sample like any other. `RdmaWorkersChargeTest.RetryIsRechargedAndStillLearns` fails and re-charges a slice and asserts the selector's inflight balance and that the release teaches the selection EWMA; `RdmaWorkersSharedQpTest.RetriedSliceIsStillCountedByItsLane` runs the failure→re-queue→re-post→success sequence through `handleCompletion` and asserts the lane count, inflight/posted bytes return to zero and the retry's completion yields the expected transmit sample.

**Relation to #3511.** It fixed the retry/fallback charge with `charged_dev_id` + `charged_bytes` + `chargeSliceQuota` (release, then re-charge). This PR keeps its own mechanism instead — `charged_dev` is exchanged for -1 by whichever path releases it, so a pooled QP's poller cannot race the reconcile, and `selectMultiPath` charges each slice its real length so there is no estimate for `charged_bytes` to correct — and follows #3511 on the two semantics it set: the selection EWMA samples a successful attempt's post→completion time, and baseline mode tracks no inflight. #3511's `quota_accounting_test.cpp` is kept with its cases re-pointed at the selector API; the slice-level ones live in `RdmaWorkersChargeTest`.

**Review by catyans — APPROVED** [review]
LGTM. Re-reviewed `ab129170`: queue-ahead now uses the posted NIC backlog, retries are recharged, and shared-QP cleanup is returned to the owning worker. The added tests cover these accounting and rate-meter cases. Thanks @SongOf and @alogfans for the thorough follow-up. Please wait for CI before merge.

**Review by alogfans — APPROVED** [review]
LGTM

**Review by he-yufeng — COMMENTED** [review]
Read through the three commits, mainly to check this against what I verified in #3750 (defects 2 and 3 on current main).

Both defects are closed at the mechanism level:

- The drop predictor no longer folds queueing into bandwidth. `DeadlineMlu` charges queueing as an additive `bytes_ahead` term over the metered wire rate, and `pickForDispatch` passes `dispatching_bytes_`, which already includes the owners picked earlier in the same call. An N-slice request no longer multiplies one shared wait by N.
- The arbitration side now orders against `getTransmitBandwidth(dev_id)` instead of the constant, with `getPostedBytes` as the queue ahead. `Step3DropCountsBytesAlreadyDispatched` pins the exact scenario from the issue: a tight owner drops behind a dispatched 1 MB, and stops dropping once that megabyte completes.

I also traced the new counter for leak paths, since `dispatching_bytes_` is only balanced if every Dispatching exit decrements: `cancel()` rejects owners already dispatching, `retireBatch` refuses non-terminal owners, and `dropOwner` only ever sees Queued owners, so `complete()` is the single exit and the decrement is symmetric with the increment (both gated on `degradation_eligible`). Same care visible in `rechargeSlice`: charging the new device before releasing the old one means a racing release pays back whichever device is on record, never both and never neither.

One thing I would like confirmed rather than changed. The two layers deliberately read different `bytes_ahead`: the drop side counts dispatched-and-not-completed (which includes slices still sitting in worker queues), while the arbitration side counts posted-to-hardware only. The net effect is that the irreversible decision is the conservative one, which is the right way around. But the comment on `DeadlineMlu` says both layers "rank the same request the same way", and with different inputs they only approximately do. Maybe worth one sentence in that header saying the drop side is intentionally the conservative reading, so nobody later "fixes" the asymmetry in the wrong direction.

**Comment by SongOf** [comment]
@he-yufeng confirmed on both.

The asymmetric `bytes_ahead` is intentional: each layer reads the bytes it can actually see ahead of the request. The drop decides before the request reaches a worker, so everything dispatched and not yet completed precedes it; the arbitration decides at posting time, when the worker-queue slices are its own contenders, so only the hardware backlog plus the slots already chosen precede a slot. The drop side therefore reads the wider set, and it is the irreversible decision — the conservatism runs the right way. I'll add a sentence to `deadline_mlu.h` saying so, so the asymmetry is not "fixed" in either direction.

`dispatching_bytes_`: your trace matches mine — `complete()` is the only `Dispatching → Terminal` exit and the only decrement, both sides gated on `degradation_eligible`. `Step3DropCountsBytesAlreadyDispatched` pins the effect but not the invariant; I'll add a unit test that checks `dispatching_bytes_` equals the eligible bytes currently in `Dispatching` after pick / complete / cancel-while-dispatching / drop / retireBatch. Both in the next push.
